php/php-rr-sdk.md

```diff
@@ -0,0 +1,692 @@
+{
+    yield new SignalHandler('name', function($input){
+    
+    });       
+}
+```
+
+> Need Temporal authors feedback.
+
+#### Handle Signals
+Incoming signals can be handled during any workflow `yield` call. Workflow should allow to explicitly wait for signal:
+
+```php
+public function run(string $input): string
+{
+    yield new SignalHandler('name', function($input){
+        // do something
+    });       
+
+    yield new WaitSignal('name');
+}
+```
+
+### Sessions
+Sessions can be created as sub-activity group, using similar approach as `withTimeout` and other methods. However,
+the explicit yield required. 
+
+```php
+public function run(string $input): string
+{
+    $session = yield $this->activities->createSession(100, 100); // timeouts
+    
+    $result = yield $session->doSomething($input)->get();
+    $result = yield $session->somethingElse($result)->get();
+    
+    yield $session->complete();
+    
+    return $result;
+}
+```
+
+### Deterministic Time
+Workflows must avoid calling SPL functions `time()` and `date()`. Context method must be used instead:
+
+```php
+$ctx->getNow(); //DateTimeImmutable object.
+```
+
+### Side Effects
+Similar to Golang SDK the side effects can be registered using yield call:
+
+```php
+$result = yield new SideEffect(function(){
+    return mt_rand(0, 1000);
+});
+```
+
+Using syntax sugar (point of discussion):
+
+```php
+$result = yield $this->sideEffect(function(){
+    return mt_rand(0, 1000);
+});
+```
+
+> Similar to other SDKs, PHP version **does not** guarantee that sideEffect won't be called again during the replay. 
+
+### Error Handling
+Like in Java version of Temporal SDK the workflow errors can be delivered in form of exceptions triggered on `yield`:
+
+```php
+public function run(string $input): string
+{
+    try {
+        return yield $this->activities->doSomething($input)->get();
+    } catch (ActivityException $e) {
+        // get related activity ID
+        $this->logger->warning(
+            "Activity %s failed with %s",
+            $e->getActivityID(),
+            $e->getMessage()
+        );       
+    }
+}
+``` 
+
+Other errors can be triggered on workflow engine level, protocol error, etc. Exceptions must be diveded into following groups:
+
+Exception Type | Parent | Comment 
+--- | --- | ---
+EngineException | - | Low level exception, usually local.
+StalledException | EngineException | Triggered to indicate the workflow processing should stop (see below).
+WorkflowException | EngineException | Generic workflow error (like activity can not be created and etc)
+TimeoutException | WorkflowException | Workflow timed out (must contain information about actual timeout type). 
+TerminateException | WorkflowException | Workflow needs to be terminated.
+ActivityException | WorkflowException | Processing errors.
+ActivityTimeoutException | ActivityException | Activity processing timeouted (must contain information about actual timeout type). 
+
```

Comment:
We recently redefined the way failures are defined in the service proto APIs to allow exception chaining across process boundaries in different languages. This deserves a separate discussion.

Reply:
I'm fine leaving it out of this proposal for now. But I have to outline that `yield` can cause the exceptions of different types.